src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,423 @@
+                    peer = i;
+                }
+                // need to wait until all ranks complete step - 1, because of double buffering
+                if (get_rank_step(task, peer, 0) < task->bcast_linear.step - 1) {
+                    // rank is not ready, lets wait
+                    return;
+                }
+            }
+            task->bcast_linear.stage = STAGE_COPY;
+            if (task->bcast_linear.step < task->bcast_linear.num_steps) {
+                // go to next iteration
+                task->bcast_linear.stage = STAGE_COPY;
+                return;
+            } else {
+                // finish
+                st = ucc_tl_cuda_shm_barrier_start(trank, task->bar);
+                if (ucc_unlikely(st != UCC_OK)) {
+                    ucc_error("failed to start barrier from root rank");
+                    task->super.status = st;
+                    return;
+                }
+                task->bcast_linear.stage = STAGE_WAIT_COMPLETION;
+            }
+        case STAGE_WAIT_COMPLETION:
+            st = ucc_tl_cuda_shm_barrier_test(trank, task->bar);
+            if (st != UCC_OK) {
+                // peers still working, lets check next time
+                task->super.status = st;
+                return;
+            }
+            // set barrier free to unlock others, this is roots responsibility
+            ucc_debug("Release bar: %p with tag: %ld", task->bar,
+                      task->bar->tag);
+            task->bar->tag = UCC_TAG_FREE;
+            ucc_tl_cuda_put_sync_root(task, task->bcast_linear.root);
+            task->super.status = UCC_OK;
+        default:
+            break;
+        }
+    } else {
+        // clients
+        // fall-through between cases is intentional
+        switch (task->bcast_linear.stage) {
+        case STAGE_WAIT_ROOT:
+            if (get_rank_step(task, task->bcast_linear.root, 0) >
+                task->bcast_linear.step) {
+                task->bcast_linear.stage = STAGE_CLIENT_COPY;
+                break;
+            } else {
+                return;
+            }
+        case STAGE_CLIENT_COPY:
+            // need to copy from root's scratch buffer
+            dbuf = PTR_OFFSET(task->bcast_linear.sbuf, offset_buff);
+            sbuf = PTR_OFFSET(TASK_SCRATCH(task, task->bcast_linear.root),
+                              task->bcast_linear.step % 2 * chunk_size);
+            st   = ecopy(dbuf, sbuf, chunk_size, exec,
+                         &task->bcast_linear.exec_task);
+            if (st != UCC_OK) {
+                ucc_error("failed to post ecopy task at client");
+                task->super.status = st;
+                return;
+            }
+            task->bcast_linear.stage = STAGE_CLIENT_COPY_WAIT;
+        case STAGE_CLIENT_COPY_WAIT:
+            etask = task->bcast_linear.exec_task;
+            ucc_assert(NULL != etask);
+            st = ucc_ee_executor_task_test(etask);
+            if (st == UCC_OK) {
```

Comment:
better to save status and return if `st != UCC_OK` and take the rest of the code out of the if/then/else body.